### PR TITLE
Feat: Add Fiber server support

### LINKED ISSUE #301: Open to new routers?
Hi, i was thinking about adding Fiber https://github.com/gofiber/fiber as a server option, but before i started wanted to check if you are open to new server options?

Thanks.

### PR DESCRIPTION
**Pull request by Jleagle**
Fixes #301 

Massive thank you to @gnirb for already doing some of the PR.

### CODE REVIEW

**Comment by gnirb** [comment]
Hello @Jleagle, how is the PR going? Can I be of any assistance?

**Comment by Jleagle** [comment]
@gnirb 
Hi, i made a start and merged in most of your work. Last time i tested it, it didn't actually generate anything 🤷. Not had much time to find out why recently. Any help would be appreciated.

**Comment by gnirb** [comment]
> @gnirb Hi, i made a start and merged in most of your work. Last time i tested it, it didn't actually generate anything 🤷. Not had much time to find out why recently. Any help would be appreciated.

Anything outstanding now before removing "draft"-status for the PR?

**Comment by Jleagle** [comment]
@deepmap-marcinr Hi, any chance one of your team could take a look at this please? Thanks!

**Comment by mfiumara** [comment]
Nice work, is it already usable from your branch? Or still requires testing?

**Comment by Jleagle** [comment]
@mfiumara 
I have not started using it in production yet but so far everything works, lots of the code comes from @gnirb's branch who uses it internally too. More testing can only be a good thing though.

File: examples/petstore-expanded/fiber/api/petstore-server.gen.go
Comment (leonklingele) [inline examples/petstore-expanded/fiber/api/petstore-server.gen.go:51]: This could use fiber's `func (c *Ctx) QueryParser(out interface{}) error` func https://docs.gofiber.io/api/ctx/#queryparser

File: examples/petstore-expanded/fiber/api/petstore-server.gen.go
Comment (leonklingele) [inline examples/petstore-expanded/fiber/api/petstore-server.gen.go:46]: This doesn't do anything.

**Comment by leonklingele** [comment]
Good job! 🥳  Added some comments.

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:26]: These funcs should call `t.Testing()` to mark them as a testing helper utility.

`Helper marks the calling function as a test helper function. When printing file and line information, that function will be skipped. Helper may be called simultaneously from multiple goroutines.`

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:38]: This should also log the `err`:

```suggestion
		t.Fatalf("Failed to test request, URL=%q: %v", rawURL, err)
```

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [reply pkg/fiber-middleware/oapi_validate_test.go:26]: Has this been resolved?! Most recent code doesn't look like it :o

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:29]:
```suggestion
		t.Fatalf("invalid URL %q: %v", rawURL, err)
```

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:47]:
```suggestion
		t.Fatalf("invalid url %q: %v", rawURL, err)
```

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:52]:
```suggestion
		t.Fatalf("failed to marshal: %v", err)
```

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:45]: Add call to `t.Testing()`

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate_test.go:62]:
```suggestion
		t.Fatalf("Failed to test request, URL=%q: %v", rawURL, err)
```

**Comment by deepmap-marcinr** [comment]
This looks good overall, with some cosmetics issues.

Can you rearrange your imports into 3 sections - stdlib, third party imports, than oapi-codegen imports? It's what other files do.

This change is isolated to Fiber mainly, but it gives us yet another router to support. I really need to think about making this modular somehow, so that we don't need to change the core oapi-codegen to provide router specific code, but this isn't anything to do with your commit. Please respond to the good questions by @leonklingele  and I'm happy to merge once everything is resolved.

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (Jleagle) [reply pkg/fiber-middleware/oapi_validate_test.go:26]: I don't have a `t.Testing()`?

**Comment by Jleagle** [comment]
> This looks good overall, with some cosmetics issues.
> 
> Can you rearrange your imports into 3 sections - stdlib, third party imports, than oapi-codegen imports? It's what other files do.
> 
> This change is isolated to Fiber mainly, but it gives us yet another router to support. I really need to think about making this modular somehow, so that we don't need to change the core oapi-codegen to provide router specific code, but this isn't anything to do with your commit. Please respond to the good questions by @leonklingele and I'm happy to merge once everything is resolved.

Imports updated.

@gnirb Would you be able to help answer [leonklingele](https://github.com/leonklingele)'s questions?

**Comment by leonklingele** [comment]
Sorry! Meant to say t.Helper()

**Comment by gnirb** [comment]
> @gnirb Would you be able to help answer [leonklingele](https://github.com/leonklingele)'s questions?

Sure, I'll take a look!

File: internal/test/issues/issue-head-digit-of-httpheader/issue.gen.go
Comment (gnirb) [reply internal/test/issues/issue-head-digit-of-httpheader/issue.gen.go:6]: Don't know, but I'll revert it

File: internal/test/strict-server/fiber/server.go
Comment (gnirb) [reply internal/test/strict-server/fiber/server.go:1]: Not sure, changing..

File: examples/petstore-expanded/fiber/api/petstore-server.gen.go
Comment (gnirb) [reply examples/petstore-expanded/fiber/api/petstore-server.gen.go:51]: After some investigation it seems like this requires all the generated request structs to be tagged with `query: <name>`. 

Gonna leave it as a future improvement, since it looks like it will require changes to the type generation potentially affecting the other server implementations too.

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (gnirb) [reply pkg/fiber-middleware/oapi_validate_test.go:49]: You mean `t.Helper()` right? That has been fixed

File: pkg/fiber-middleware/oapi_validate_test.go
Comment (gnirb) [reply pkg/fiber-middleware/oapi_validate_test.go:26]: You mean t.Helper() right? That has been fixed

**Comment by gnirb** [comment]
Alright @leonklingele comments addressed, please take another look if this feels good to go

**Comment by Jleagle** [comment]
> This looks good overall, with some cosmetics issues.
> 
> Can you rearrange your imports into 3 sections - stdlib, third party imports, than oapi-codegen imports? It's what other files do.
> 
> This change is isolated to Fiber mainly, but it gives us yet another router to support. I really need to think about making this modular somehow, so that we don't need to change the core oapi-codegen to provide router specific code, but this isn't anything to do with your commit. Please respond to the good questions by @leonklingele and I'm happy to merge once everything is resolved.

@deepmap-marcinr. I think everything has been resolved 👍

File: go.mod
Comment (leonklingele) [inline go.mod:8]: The `adaptor` middleware now is part of fiber stdlib: https://github.com/gofiber/fiber/tree/77c1b4868a686254a839fbbb88d7f46cbadb055a/middleware/adaptor

File: go.mod
Comment (leonklingele) [inline go.mod:9]: Update this dependency.

File: go.mod
Comment (leonklingele) [inline go.mod:17]: Update this dependency.

File: pkg/codegen/templates/imports.tmpl
Comment (leonklingele) [inline pkg/codegen/templates/imports.tmpl:31]: Move this up one line so this stays lexically sorted.

File: internal/test/strict-server/fiber/server.gen.go
Comment (leonklingele) [inline internal/test/strict-server/fiber/server.gen.go:325]: This should call `ctx.WriteString` for performance reasons.

File: internal/test/strict-server/fiber/server.gen.go
Comment (leonklingele) [inline internal/test/strict-server/fiber/server.gen.go:366]: This already is a string and the type conversion can be dropped.

File: internal/test/strict-server/fiber/server.gen.go
Comment (leonklingele) [inline internal/test/strict-server/fiber/server.gen.go:391]: This already is a string and the type conversion can be dropped.

File: internal/test/strict-server/fiber/server.gen.go
Comment (leonklingele) [inline internal/test/strict-server/fiber/server.gen.go:444]: This already is a string and the type conversion can be dropped.

File: internal/test/strict-server/fiber/server.gen.go
Comment (leonklingele) [inline internal/test/strict-server/fiber/server.gen.go:585]: This should call `ctx.WriteString` for performance reasons.

**Review by leonklingele — COMMENTED** [review]
Last-minute review incoming 🫨

Almost perfect, just some small modifications which also make it a bit faster by removing unnecessary type conversions. Thanks for your work here, @Jleagle!

**Comment by leonklingele** [comment]
@Jleagle please also do one final rebase to `master` :)

File: pkg/fiber-middleware/oapi_validate.go
Comment (leonklingele) [inline pkg/fiber-middleware/oapi_validate.go:96]: I've opened https://github.com/gofiber/fiber/pull/2461 so this can be done with fiber stdlib — not requiring to use other modules.

File: go.mod
Comment (Jleagle) [reply go.mod:8]: I dont think this has been versioned yet

Comment (leonklingele) [reply go.mod]: fiber v2.46.0 just dropped 🥳

**Review by leonklingele — APPROVED** [review]
lgtm! @deepmap-marcinr what do you say? :)

**Comment by deepmap-marcinr** [comment]
> lgtm! @deepmap-marcinr what do you say? :)

This is a big change and another router to maintain, but it's nicely isolated. Could you rebase and fix the conflicts, and then I'll merge it, and release it in 1.13.0, which I intend to release as soon as this commit is ready.

**Comment by deepmap-marcinr** [comment]
I pulled your change to test locally, and some things are missing.

1) run `go mod tidy`
2) regenerate all your fiber boilerplate
3) Tests fail:
```
# github.com/deepmap/oapi-codegen/internal/test/strict-server/fiber
internal/test/strict-server/fiber/server.gen.go:810:10: request.PType undefined (type ReservedGoKeywordParametersRequestObject has no field or method PType)
```

File: pkg/codegen/templates/strict/strict-fiber.tmpl
Comment (deepmap-marcinr) [inline pkg/codegen/templates/strict/strict-fiber.tmpl:22]: This needs to be:

`request.{{.GoName}} = {{.GoVariableName}}`

File: internal/test/strict-server/fiber/server.go
Comment (deepmap-marcinr) [inline internal/test/strict-server/fiber/server.go:102]: You are missing a function here.

```
func (s StrictServer) ReservedGoKeywordParameters(ctx context.Context, request ReservedGoKeywordParametersRequestObject) (ReservedGoKeywordParametersResponseObject, error) {
	return ReservedGoKeywordParameters200TextResponse(""), nil
}

```

**Comment by deepmap-marcinr** [comment]
@Jleagle  - please see the PR I sent you which fixes these problems.
https://github.com/Jleagle/oapi-codegen/pull/31

After you merge that, pls make lint pass.

**Pull request merged**
